core/circuit_breaker/circuit_breaker.go

```diff
@@ -2,252 +2,520 @@ package circuit_breaker
 
 import (
 	"sync/atomic"
-	"time"
+	"unsafe"
 
 	"github.com/alibaba/sentinel-golang/core/base"
-	"github.com/alibaba/sentinel-golang/core/stat"
+	sbase "github.com/alibaba/sentinel-golang/core/stat/base"
 	"github.com/alibaba/sentinel-golang/util"
 )
 
+type Status int32
+
+/**
+  Circuit Breaker Status Machine:
+
+                                 switch to open based on rule
+         +-----------------------------------------------------------------------+
+         |                                                                       |
+         |                                                                       v
++----------------+                   +----------------+      Probe      +----------------+
+|                |                   |                |<----------------|                |
+|                |   Probe succeed   |                |                 |                |
+|     Closed     |<------------------|    HalfOpen    |                 |      Open      |
+|                |                   |                |   Probe failed  |                |
+|                |                   |                +---------------->|                |
++----------------+                   +----------------+                 +----------------+
+*/
+const (
+	Closed Status = iota
+	HalfOpen
+	Open
+)
+
+func (s *Status) get() Status {
+	statusPtr := (*int32)(unsafe.Pointer(s))
+	return Status(atomic.LoadInt32(statusPtr))
+}
+func (s *Status) set(update Status) {
+	statusPtr := (*int32)(unsafe.Pointer(s))
+	newStatus := int32(update)
+	atomic.StoreInt32(statusPtr, newStatus)
+}
+func (s *Status) casStatus(expect Status, update Status) bool {
+	statusPtr := (*int32)(unsafe.Pointer(s))
+	oldStatus := int32(expect)
+	newStatus := int32(update)
+	return atomic.CompareAndSwapInt32(statusPtr, oldStatus, newStatus)
+}
+
+// StatusSwitchListener could listen on circuit breaker status switch
+type StatusSwitchListener interface {
+	onSwitchToClosed(prev Status, rule Rule)
+
+	onSwitchToOpen(prev Status, rule Rule)
```

Comment:
We could also carry the triggered "snapshot" value when the circuit breaker transforms to open.